data/src/main/java/com/linkedin/data/Data.java

```diff
@@ -693,6 +720,29 @@ private static boolean reachable(DataComplex source, Object destination)
     return false;
   }
 
+  /**
+   * Check for cycles.
+   *
+   * <p>This checks if the given {@link DataComplex} is present in the given set of ancestor complexes. If yes,
+   * then it indicates that a cycle exists and we throw an exception. If not, we add it to the set.</p>
+   *
+   * @param dataComplex The current {@link DataComplex}
+   * @param ancestorSet The set containing all ancestors of the {@link DataComplex}
+   * @param pathList The list of paths encountered so far. Used to populate a useful exception message when a cycle is
+   *                 detected.
+   * @throws IOException If this {@link DataComplex} is present in the ancestor set indicating a cycle.
+   */
+  private static void checkForCyclesAndAdd(DataComplex dataComplex,
```

Comment:
Why bother having the "add" logic bundled in with this method? I can think of a couple reasons to move it out:
- Improves readability in the containing method because there's visible symmetry between the "add ancestor" and "remove ancestor" statements.
- Adds an on-success side effect to an otherwise assertion-like method. Method is cleaner and simpler if it's just a "check for cycles" method.
- Adding the ancestor isn't really related to this method and could just be included outside the method. Could just move the last line outside to get the same effect.

Reply:
Fixed